kernel/kernel-api/src/main/java/io/delta/kernel/internal/columndefaults/ColumnDefaults.java

```diff
@@ -0,0 +1,185 @@
+ * data to the table is not allowed.
+ */
+public class ColumnDefaults {
+
+  private static final String DEFAULT_VALUE_METADATA_KEY = "CURRENT_DEFAULT";
+
+  /** Don't allow data writes to tables with default values */
+  public static void blockWriteIfEnabled(Row transactionState) {
+    if (extractFieldsWithDefaultValues(TransactionStateRow.getLogicalSchema(transactionState))
+        .findAny()
+        .isPresent()) {
+      throw new UnsupportedOperationException(
+          "Writing with Column Default values is not supported yet.");
+    }
+  }
+
+  /**
+   * Validate Column Default values in the provided metadata. Kernel only supports literal default
+   * values. See {validateLiteral}.
+   *
+   * @param schema target table schema
+   * @param isEnabled When the feature is disabled, no column default is allowed.
+   * @param isIcebergCompatV3Enabled Kernel currently requires IcebergCompatV3 to be enabled when
+   *     using Column Defaults
+   * @throws KernelException when the table contains invalid default value
+   */
+  public static void validateSchema(
+      StructType schema, boolean isEnabled, boolean isIcebergCompatV3Enabled) {
+    if (isEnabled && !isIcebergCompatV3Enabled) {
+      throw new KernelException(
+          "In Delta Kernel, default values table feature requires IcebergCompatV3 to be enabled.");
+    }
+    Stream<StructField> defaultValues = extractFieldsWithDefaultValues(schema);
+    if (!isEnabled) {
+      if (defaultValues.findAny().isPresent()) {
+        throw new KernelException(
+            "This table does not enable table features for setting column defaults");
+      }
+    } else {
+      // This check will be relaxed once kernel supports default values with expression
+      defaultValues.forEach(
+          field -> {
+            String defaultValue = getRawDefaultValue(field);
+            try {
+              validateLiteral(field.getDataType(), defaultValue);
+            } catch (DateTimeParseException
+                | IllegalArgumentException
+                | UnsupportedOperationException e) {
+              throw DeltaErrors.nonLiteralDefaultValue(defaultValue);
+            }
+          });
+    }
+  }
+
+  /**
+   * Validate that the schema only contains literal default values as a requirement of
+   * IcebergCompat.
+   *
+   * @param schema table schema
+   */
+  public static void validateSchemaForIcebergCompat(StructType schema, String compatVersion) {
+    extractFieldsWithDefaultValues(schema)
+        .forEach(
+            field -> {
+              String defaultValue = getRawDefaultValue(field);
+              try {
+                validateLiteral(field.getDataType(), defaultValue);
+              } catch (DateTimeParseException
+                  | IllegalArgumentException
+                  | UnsupportedOperationException e) {
+                throw DeltaErrors.icebergCompatRequiresLiteralDefaultValue(
```

Comment:
I think my comment before was related to what exception we throw. For example, if we find a column default for a data type we don't support defaults for, we should throw that exception, not the required literalValue

Reply:
Can we make `validateLiteral` have a very clear defined docs + throws behavior?

@throws UnsupportedOperationException if Kernel doesn't support default values for the data type
@throws IllegalArgumentException if the value is not a literal value

--> In `validateLiteral` we should only throw those exceptions (so need to catch the parse exceptions there)
--> here, we can be specific about our behavior, i.e. take IllegalArgumentException and change to icebergCompatRequiresLiteralDefaultValue; probably just let UnsupportedOperationException fail upward